### PR TITLE
All Model Created

### LINKED ISSUE #2: Create Seed for **Librarian** and Migrate Database

### PR DESCRIPTION
**Pull request by raj-khan**
All Models are created under this PR.

### CODE REVIEW

File: app/Models/Author.php [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Author extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/Author.php]: Use 'SoftDeletes' trait.

File: app/Models/Book.php [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Book extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/Book.php]: Use 'SoftDeletes' trait.

File: app/Models/AuthorBook.php [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class AuthorBook extends Model
```
Comment (me-shaon) [inline app/Models/AuthorBook.php]: This class should extend 'Pivot' class not 'Model' class.

File: app/Models/BookCopy.php [hunk 4]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class BookCopy extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/BookCopy.php]: Use 'SoftDeletes' trait.

File: app/Models/BookGenre.php [hunk 5]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class BookGenre extends Model
```
Comment (me-shaon) [inline app/Models/BookGenre.php]: This class should extend 'Pivot' class not 'Model' class.

File: app/Models/BookUser.php [hunk 6]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class BookUser extends Model
```
Comment (me-shaon) [inline app/Models/BookUser.php]: This class should extend 'Pivot' class not 'Model' class.

File: app/Models/Genre.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Genre extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/Genre.php]: Use 'SoftDeletes' trait.

File: app/Models/LoanRequest.php [hunk 8]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class LoanRequest extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/LoanRequest.php]: Use 'SoftDeletes' trait.

File: app/Models/Publication.php [hunk 9]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class Publication extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/Publication.php]: Use 'SoftDeletes' trait.

File: app/Models/ReturnRequest.php [hunk 10]
```diff
@@ -0,0 +1,11 @@
+<?php
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class ReturnRequest extends Model
+{
+    use HasFactory;
```
Comment (me-shaon) [inline app/Models/ReturnRequest.php]: Use 'SoftDeletes' trait.

File: app/Models/User.php [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+namespace App\Models;
+
+use Illuminate\Contracts\Auth\MustVerifyEmail;
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Foundation\Auth\User as Authenticatable;
+use Illuminate\Notifications\Notifiable;
+
+class User extends Authenticatable
+{
+    use HasFactory, Notifiable;
```
Comment (me-shaon) [inline app/Models/User.php]: Use 'SoftDeletes' trait.

File: database/factories/AuthorBookFactory.php [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+    /**
+     * Define the model's default state.
+     *
+     * @return array
+     */
+    public function definition()
+    {
+        return [
+            //
```
Comment (me-shaon) [inline database/factories/AuthorBookFactory.php]: You should finish this method implementation.  
Same for other factory classes.

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+class CreateUsersTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('users', function (Blueprint $table) {
```
Comment (me-shaon) [inline database/migrations/2014_10_12_000000_create_users_table.php:17]: Where are 'phone', 'role', 'is_banned' etc. fields?

File: database/migrations/2020_10_30_182222_create_publications_table.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+class CreatePublicationsTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('publications', function (Blueprint $table) {
```
Comment (me-shaon) [inline database/migrations/2020_10_30_182222_create_publications_table.php:16]: Please update all the migrations with their fields.

**Review by me-shaon — CHANGES_REQUESTED** [review]

File: database/migrations/2014_10_12_000000_create_users_table.php (hunk 13 above)
Comment (raj-khan) [reply database/migrations/2014_10_12_000000_create_users_table.php:17]: It's the default from the installation file, I will start working on migration files.

File: app/Models/Book.php (hunk 2 above)
Comment (raj-khan) [reply app/Models/Book.php]: Done.

File: app/Models/AuthorBook.php (hunk 3 above)
Comment (raj-khan) [reply app/Models/AuthorBook.php]: Done.

File: app/Models/BookCopy.php (hunk 4 above)
Comment (raj-khan) [reply app/Models/BookCopy.php]: Done.

File: app/Models/BookGenre.php (hunk 5 above)
Comment (raj-khan) [reply app/Models/BookGenre.php]: Done.

File: app/Models/BookUser.php (hunk 6 above)
Comment (raj-khan) [reply app/Models/BookUser.php]: Done.

File: app/Models/Genre.php (hunk 7 above)
Comment (raj-khan) [reply app/Models/Genre.php]: Done.

File: app/Models/LoanRequest.php (hunk 8 above)
Comment (raj-khan) [reply app/Models/LoanRequest.php]: Done.

File: app/Models/Publication.php (hunk 9 above)
Comment (raj-khan) [reply app/Models/Publication.php]: Done.

File: app/Models/ReturnRequest.php (hunk 10 above)
Comment (raj-khan) [reply app/Models/ReturnRequest.php]: Done.

File: app/Models/User.php (hunk 11 above)
Comment (raj-khan) [reply app/Models/User.php]: Done.

File: database/factories/AuthorBookFactory.php (hunk 12 above)
Comment (raj-khan) [reply database/factories/AuthorBookFactory.php]: Done.

File: database/migrations/2020_10_30_182222_create_publications_table.php (hunk 14 above)
Comment (raj-khan) [reply database/migrations/2020_10_30_182222_create_publications_table.php:16]: Done.

File: app/Models/Author.php [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+class Author extends Model
+{
+    use HasFactory;
+    use SoftDeletes;
```
Comment (me-shaon) [inline app/Models/Author.php]: No need to use a separate 'use' statement. You can simply write them in the same line, like: 
```
use HasFactory, SoftDeletes;
```

Comment (me-shaon) [reply app/Models/Author.php]: Fix in the other places too.

File: app/Models/AuthorBook.php [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Relations\Pivot;
+
+class AuthorBook extends Pivot
+{
+    use HasFactory;
+
```
Comment (me-shaon) [inline app/Models/AuthorBook.php:11]: Use only one single empty line.

Comment (me-shaon) [reply app/Models/AuthorBook.php:11]: Fix in other places too.

File: app/Models/AuthorBook.php [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+
+    public function author(){
```
Comment (me-shaon) [inline app/Models/AuthorBook.php]: The `{` should be on the next line.

Comment (me-shaon) [reply app/Models/AuthorBook.php]: Fix in other places too.

File: database/migrations/2020_10_30_182601_create_book_users_table.php [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+class CreateBookUsersTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('book_users', function (Blueprint $table) {
```
Comment (me-shaon) [inline database/migrations/2020_10_30_182601_create_book_users_table.php]: The table name was 'book_user'.

File: database/migrations/2020_10_30_182601_create_book_users_table.php [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+    }
+
+    /**
+     * Reverse the migrations.
+     *
+     * @return void
+     */
+    public function down()
+    {
+        Schema::dropIfExists('book_users');
```
Comment (me-shaon) [inline database/migrations/2020_10_30_182601_create_book_users_table.php]: You need to drop the 'foreign key' fields first, then drop the table.

Comment (me-shaon) [reply database/migrations/2020_10_30_182601_create_book_users_table.php]: Fix it in other places too.

**Review by me-shaon — CHANGES_REQUESTED** [review]

File: app/Models/Author.php (hunk 15 above)
Comment (raj-khan) [reply app/Models/Author.php]: Fixed

File: app/Models/AuthorBook.php (hunk 17 above)
Comment (raj-khan) [reply app/Models/AuthorBook.php]: Fixed

File: database/migrations/2020_10_30_182601_create_book_users_table.php (hunk 18 above)
Comment (raj-khan) [reply database/migrations/2020_10_30_182601_create_book_users_table.php]: Renamed.

File: database/migrations/2020_10_30_182601_create_book_users_table.php (hunk 19 above)
Comment (raj-khan) [reply database/migrations/2020_10_30_182601_create_book_users_table.php]: Fixed.

File: app/Models/AuthorBook.php (hunk 16 above)
Comment (raj-khan) [reply app/Models/AuthorBook.php:11]: Fixed.

File: app/Models/Author.php [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+class Author extends Model
+{
+    use HasFactory, SoftDeletes;
+
+    public function authors()
```
Comment (me-shaon) [inline app/Models/Author.php]: The relationship seems wrongly named. Please check.

File: app/Models/Book.php [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+class Book extends Model
+{
+    use HasFactory, SoftDeletes;
+
+    public function publication()
+    {
+        return $this->belongsTo(Publication::class);
+    }
+
+    public function bookCopies()
```
Comment (me-shaon) [inline app/Models/Book.php]: Name the relationship 'copies'.

File: app/Models/Book.php [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return $this->hasMany(BookCopy::class);
+    }
+
+    public function bookGenres()
```
Comment (me-shaon) [inline app/Models/Book.php]: Name the relationship 'genres'.

File: app/Models/Book.php [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        return $this->hasMany(BookGenre::class);
+    }
+
+    public function authorBooks()
```
Comment (me-shaon) [inline app/Models/Book.php]: Is this relationship needed here?

File: database/migrations/2014_10_12_000000_create_users_table.php [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+            $table->id();
+            $table->string('name');
+            $table->string('password');
+            $table->string('email')->unique();
+            $table->string('phone');
+            $table->string('role')->default('user')->comment('librarian/user');
```
Comment (me-shaon) [inline database/migrations/2014_10_12_000000_create_users_table.php]: Use 'Constants' to define the roles.

**Review by me-shaon — CHANGES_REQUESTED** [review]

File: app/Models/Author.php (hunk 20 above)
Comment (raj-khan) [reply app/Models/Author.php]: Corrected.

File: app/Models/Book.php (hunk 21 above)
Comment (raj-khan) [reply app/Models/Book.php]: Changed.

File: app/Models/Book.php (hunk 22 above)
Comment (raj-khan) [reply app/Models/Book.php]: changed.

File: app/Models/Book.php (hunk 23 above)
Comment (raj-khan) [reply app/Models/Book.php]: Yes, I understand, & it's removed.

File: database/migrations/2014_10_12_000000_create_users_table.php (hunk 24 above)
Comment (raj-khan) [reply database/migrations/2014_10_12_000000_create_users_table.php]: Used.

File: app/Models/Author.php [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\SoftDeletes;
+
+class Author extends Model
+{
+    use HasFactory, SoftDeletes;
+
+    public function authorBooks()
```
Comment (me-shaon) [inline app/Models/Author.php:13]: It should be 'books', rather than 'authorBooks', because the Model is already representing 'Author'.

**Review by me-shaon — APPROVED** [review]
Minor change commented. I'm merging the PR, but keep in mind to fix it in the next PR.

**Pull request merged**